libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs

```diff
@@ -238,62 +272,127 @@ propFilters ckFilter hkFilter statusFilter nes = do
+      let comMembers = Map.keysSet cm
+      let comStateMembers = Map.keysSet comStateMembers'
+
+      filterNext ToBeEnacted noFilterResult
+        `shouldSatisfy` (\res -> Map.keysSet res == nextComMembers `Set.difference` comMembers)
+
+      filterNext ToBeRemoved noFilterResult
+        `shouldSatisfy` (\res -> Map.keysSet res == (comMembers `Set.union` comStateMembers) `Set.difference` nextComMembers)
+
+      -- members who are both in current and nextCommittee are either ToBeExpired or NoChangeExpected
+      Map.keysSet (filterNext NoChangeExpected noFilterResult)
+        `Set.union` Map.keysSet (filterNext ToBeExpired noFilterResult)
+        `shouldSatisfy` (== (comMembers `Set.intersection` nextComMembers))
+
+      let currentEpoch = csEpochNo noFilterResult
+      cmsExpiration
+        <$> filterNext ToBeExpired noFilterResult
+          `shouldSatisfy` all (all (< currentEpoch + 1))
+
+      cmsExpiration
+        <$> filterNext NoChangeExpected noFilterResult
+          `shouldSatisfy` all (all (>= currentEpoch + 1))
+  where
+    filterNext nextEpochChange cms =
+      Map.filter
+        ( \case
+            CommitteeMemberState _ _ _ nextEpochChange' ->
+              nextEpochChange == nextEpochChange'
+        )
+        (csCommittee cms)
+
+genCommittee ::
+  forall era.
+  Era era =>
+  Gen (Maybe (Committee era))
+genCommittee = frequency [(1, pure Nothing), (9, Just <$> arbitrary)]
+
+genRelevantCommitteeState ::
+  forall era.
+  EraTxOut era =>
+  Maybe (Committee era) ->
+  Maybe (Committee era) ->
   Gen (CommitteeState era)
-genCommonMembers comMembers comStateMembers = do
-  s <- choose (0, Map.size comMembers)
-  let chosen = Map.take s comMembers
-  let x =
+genRelevantCommitteeState maybeCm nextCom = do
+  CommitteeState comStateMembers <- arbitrary @(CommitteeState era)
+  let comMembers = Map.keysSet (foldMap' committeeMembers maybeCm)
+      nextComMembers = flatColdCreds nextCom
+      third = Set.size comMembers `div` 3
+      chosen = Set.toList $ Set.take third comMembers
+      nextChosen = Set.toList $ Set.take third nextComMembers
+      x =
         Map.fromList $
           zipWith
-            (\(k1, _) (_, v2) -> (k1, v2))
-            (Map.assocs chosen)
+            (\k1 (_, v2) -> (k1, v2))
+            (chosen <> nextChosen)
             (Map.assocs comStateMembers)
-  pure $ CommitteeState $ x `Map.union` (Map.drop s comStateMembers)
+  pure $ CommitteeState $ x `Map.union` Map.drop (length chosen + length nextChosen) comStateMembers
+
+genNextCommittee ::
+  forall era.
+  EraTxOut era =>
+  Maybe (Committee era) ->
+  Gen (Maybe (Committee era))
+genNextCommittee maybeCm = do
+  let comMembers = foldMap' committeeMembers maybeCm
+  new <- arbitrary @(Map.Map (Credential 'ColdCommitteeRole (EraCrypto era)) EpochNo)
+  q <- arbitrary
+  let half = Map.size new `div` 2
+  oneof $
+    pure
+      <$> [Nothing, Just $ Committee (Map.union (Map.take half comMembers) (Map.take half new)) q]
```

Comment:
I was trying to keep the same randomly generated committee size, because I thought the sizes are generated with sense and structure.  But happy to go with your suggestion, since I agree it's less "wtf-inducing".

Reply:
To be honest with you, I am not sure I understand why you set this goal: "I was trying to keep the same randomly generated committee size". Trying to ensure that the size matches the random size does not make much sense to me. 

Moreover, here is an example run of you generator:

```haskell
comMembers = [x,y]
new = [a,b,c,d,e,f,g,h,j]
result = [x,y,a,b,c,d]
```

Why generate `[e,f,g,h,j]` just to discard it.

Also just realized it should be switched to:
```haskell
genNextCommittee maybeCm = 
  oneof [pure Nothing, Just <$> genCom]
  where
    genCom = do
        let comMembers = foldMap' committeeMembers maybeCm
        new <- arbitrary @(Map.Map (Credential 'ColdCommitteeRole (EraCrypto era)) EpochNo)
        ..
```

because there is no point in generating stuff if it will be discarded with Nothing case anyways

Reply:
I have applied this idea here and to the rest of similar functions (generating relevant committeeState and filters): 
https://github.com/input-output-hk/cardano-ledger/pull/3878/commits/e77d5210b0a4819f34646ec341d97e7a32ea7a3a